### PR TITLE
WebDriver: Let's run some JavaScript!

### PR DESCRIPTION
**Pull request by linusg**
[Screencast from 2022-11-02 18-52-32.webm](https://user-images.githubusercontent.com/19366641/199580239-61380c93-46dd-4fb7-8c2c-0b3845422302.webm)

Ref #15551

### CODE REVIEW

File: Userland/Libraries/LibWeb/CMakeLists.txt [hunk 1]
```diff
@@ -437,6 +437,7 @@ set(SOURCES
     WebAssembly/WebAssemblyTableConstructor.cpp
```
Comment (FireFox317) [inline Userland/Libraries/LibWeb/CMakeLists.txt:437]: Typo in commit message:

This ~~is~~ cannot be done

Comment (linusg) [reply Userland/Libraries/LibWeb/CMakeLists.txt:437]: Thanks, fixed!

File: Userland/Libraries/LibWeb/WebDriver/ExecuteScript.cpp [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+        return JsonValue {};
+    }
+
+    // -> type Boolean
+    // -> type Number
+    // -> type String
+    if (value.is_boolean() || value.is_number() || value.is_string()) {
+        // Success with data value.
+        auto json_string = MUST(JS::JSONObject::stringify_impl(vm, value, JS::js_undefined(), JS::js_undefined()));
+        return MUST(JsonValue::from_string(json_string));
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/WebDriver/ExecuteScript.cpp:83]: `Value` -> `String` -> `JsonValue` seems a bit excessive? At least for bools and numbers

File: Userland/Libraries/LibWeb/WebDriver/ExecuteScript.cpp [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+    // FIXME: - instance of shadow root
+    // FIXME: - a WindowProxy object
+
+    // -> has an own property named "toJSON" that is a Function
+    auto to_json = value.as_object().get_without_side_effects(vm.names.toJSON);
+    if (to_json.is_function()) {
+        // Return success with the value returned by Function.[[Call]](toJSON) with value as the this value.
+        auto value_or_error = to_json.as_function().internal_call(value, JS::MarkedVector<JS::Value> { vm.heap() });
+        if (value_or_error.is_throw_completion() || !value_or_error.value().is_string())
+            return ExecuteScriptResultType::JavaScriptError;
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/WebDriver/ExecuteScript.cpp:102]: There's a *lot* of manual `is_throw_completion` checking in this, do we want a macro at the top of the file to simplify this?
```c++
#define TRY_JS(expression)                                   \
    ({                                                       \
        auto _temporary_result = (expression);               \
        if (_temporary_result.is_error()) [[unlikely]]       \
            return ExecuteScriptResultType::JavaScriptError; \
        _temporary_result.release_value();                   \
    })
```

Or to return ExecuteScriptResult:
```c++
#define TRY_JS(expression)                                                                          \
    ({                                                                                              \
        auto _temporary_result = (expression);                                                      \
        if (_temporary_result.is_error()) [[unlikely]]                                              \
            return { ExecuteScriptResultType::JavaScriptError, _temporary_result.release_error() }; \
        _temporary_result.release_value();                                                          \
    })
```

Comment (trflynn89) [reply Userland/Libraries/LibWeb/WebDriver/ExecuteScript.cpp:102]: That may also get rid of some of the `MUST`s, assuming any of them are actually fallible

File: Userland/Services/WebDriver/Session.cpp [hunk 4]
```diff
@@ -899,6 +900,78 @@ ErrorOr<JsonValue, WebDriverError> Session::get_element_tag_name(JsonValue const
...
+        return WebDriverError::from_code(ErrorCode::ScriptTimeoutError, "Script timed out");
+    // 7. Upon fulfillment of promise with value v, let result be a JSON clone of v, and return success with data result.
+    case Web::WebDriver::ExecuteScriptResultType::PromiseResolved:
+        // NOTE: This is assumed to be a valid JSON value.
+        return MUST(JsonValue::from_string(result.json_result()));
+    // 8. Upon rejection of promise with reason r, let result be a JSON clone of r, and return error with error code javascript error and data result.
+    // FIXME: The definition of 'error' doesn't mention how to return data!
+    case Web::WebDriver::ExecuteScriptResultType::PromiseRejected:
+    case Web::WebDriver::ExecuteScriptResultType::JavaScriptError:
+        return WebDriverError::from_code(ErrorCode::JavascriptError, "Script returned an error");
```
Comment (trflynn89) [inline Userland/Services/WebDriver/Session.cpp:969]: Does `json_result` have anything useful here if we get an error? If so, we can:
```c++
return WebDriverError::from_code(ErrorCode::JavascriptError, String::formatted("Script returned an error: {}", json_result));
```

File: Userland/Libraries/LibWeb/WebDriver/ExecuteScript.cpp (hunk 2 above)
Comment (linusg) [reply Userland/Libraries/LibWeb/WebDriver/ExecuteScript.cpp]: Fair enough, it's `Value` -> `JsonValue` now :)

File: Userland/Services/WebDriver/Session.cpp (hunk 4 above)
Comment (linusg) [reply Userland/Services/WebDriver/Session.cpp]: Resolved the FIXME instead to do it properly (as a `"data"` key of the returned error JSON object)

File: Userland/Libraries/LibWeb/WebDriver/ExecuteScript.cpp (hunk 3 above)
Comment (linusg) [reply Userland/Libraries/LibWeb/WebDriver/ExecuteScript.cpp]: Didn't get rid of all manual checks, but it's an improvement. The `MUST`s were already infallible.

**Review by trflynn89 — APPROVED** [review]
LGTM, you've got some build errors though

**Comment by linusg** [comment]
Yep, just fixed those - botched attempt at making something nicer with a `Variant` :sweat_smile:

**Pull request merged**